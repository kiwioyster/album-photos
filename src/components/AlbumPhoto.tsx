import React, { useState } from 'react';
import { IPhoto } from '../type';

interface Props {
  photos: IPhoto[];
}

const AlbumPhoto: React.FC<Props> = ({ photos }) => {
  const [shownPhoto, setShownPhoto] = useState<number>();
  const clickThumbnail = (id: number) => {
    setShownPhoto(id);
  };
  return (
    <>
      {photos?.map((photo) => {
        return (
          <div key={photo.id}>
            <img
              className='album-thumbnail'
              src={photo.thumbnailUrl}
              onClick={() => clickThumbnail(photo.id)}
              alt='album thumbnail'
            ></img>
            {(() => {
              if (shownPhoto === photo.id) {
                return (
                  <img
                    className='album-photo-fullsize'
                    alt='album fullsize'
                    src={photos.find((photo) => photo.id === shownPhoto)?.url}
                  ></img>
                );
              }
            })()}
          </div>
        );
      })}
    </>
  );
};

export default AlbumPhoto;
