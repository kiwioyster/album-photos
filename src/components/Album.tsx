import React, { useState } from 'react';
import axios from 'axios';
import AlbumPhoto from './AlbumPhoto';
import { IAlbum, IUser, IPhoto } from '../type';

interface Props {
  album: IAlbum;
  users: IUser[];
}

const Album: React.FC<Props> = ({ album, users }) => {
  const [photos, setPhotos] = useState<IPhoto[]>([]);
  const PHOTO_URL = 'http://jsonplaceholder.typicode.com/photos';
  const fetchPhoto = async (albumId: number) => {
    try {
      const response = await axios.get<IPhoto[]>(
        `${PHOTO_URL}?albumId=${albumId}`
      );
      setPhotos(response.data);
    } catch (e) {
      console.error(e);
    }
  };
  const clickedAlbum = (id: number) => {
    fetchPhoto(id);
  };
  return (
    <div className='album' onClick={() => clickedAlbum(album.id)}>
      <div className='album-title'>Title: {album.title}</div>
      <div className='album-user'>
        User: {users.find((user) => user.id === album.userId)?.name}
      </div>
      <AlbumPhoto photos={photos} />
    </div>
  );
};

export default Album;
