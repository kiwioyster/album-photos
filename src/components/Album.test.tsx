import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import Album from './Album';
import Adapter from 'enzyme-adapter-react-16';
import { IAlbum, IUser, IPhoto } from '../type';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { act, cleanup } from '@testing-library/react';
import AlbumPhoto from './AlbumPhoto';

const mockAlbum: IAlbum = {
  userId: 2,
  id: 1,
  title: 'mock album',
};

const mockUsers: IUser[] = [
  {
    id: 1,
    name: 'mock user',
    username: 'mock username',
    address: 'mock address',
    phone: 'mock phone',
    website: 'mock website',
    company: 'mock company',
  },
  {
    id: 2,
    name: 'mock user 2',
    username: 'mock username 2',
    address: 'mock address 2',
    phone: 'mock phone 2',
    website: 'mock website 2',
    company: 'mock company 2',
  },
];

configure({ adapter: new Adapter() });
describe('Album', () => {
  afterEach(cleanup);
  it('should get from api', async () => {
    const mock = new MockAdapter(axios);
    const mockData: IPhoto[] = [
      {
        albumId: 1,
        id: 2,
        title: 'mock photo title',
        url: 'mock photo url',
        thumbnailUrl: 'mock thumbnail url',
      },
    ];
    mock
      .onGet('http://jsonplaceholder.typicode.com/photos?albumId=1')
      .reply(200, mockData);
    const component = mount(<Album album={mockAlbum} users={mockUsers} />);
    component.find('.album').at(0).simulate('click');
    await act(
      () =>
        new Promise<void>((resolve) => {
          setImmediate(() => {
            component.update();
            resolve();
          });
        })
    );
    expect(component.find('.album-thumbnail').length).toEqual(1);
    expect(component.find(AlbumPhoto).prop('photos')).toEqual(mockData);
  });
  it('should display correct album title and user', () => {
    const component = shallow(<Album album={mockAlbum} users={mockUsers} />);
    expect(component.find('.album-title').html()).toContain(
      'Title: mock album'
    );
    expect(component.find('.album-user').html()).toContain('User: mock user 2');
  });
});
