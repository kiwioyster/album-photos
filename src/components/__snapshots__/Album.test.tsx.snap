// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album should render 1`] = `
Array [
  <div
    className="album"
    onClick={[Function]}
  >
    <div
      className="album-title"
    >
      Title: 
      mock album
    </div>
    <div
      className="album-user"
    >
      User: 
      mock user 2
    </div>
    <AlbumPhoto
      photos={Array []}
    />
  </div>,
]
`;
